sdk/communication/azure-communication-phonenumbers/src/main/java/com/azure/communication/phonenumbers/PhoneNumbersAsyncClient.java

```diff
@@ -129,12 +138,90 @@
      *
      * @param countryCode The ISO 3166-2 country code.
      * @param searchRequest {@link PhoneNumberSearchRequest} specifying the search request
+     * @param pollInterval The time our long running operation will keep on polling
+     * until it gets a result from the server
      * @return A {@link PollerFlux} object with the reservation result
      */
     @ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
-    public PollerFlux<PollResult<PhoneNumberSearchResult>, PhoneNumberSearchResult> beginSearchAvailablePhoneNumbers(
-        String countryCode, PhoneNumberSearchRequest searchRequest) {
-        return null;
+    public PollerFlux<PhoneNumberOperationResult, PhoneNumberSearchResult> beginSearchAvailablePhoneNumbers(
+        String countryCode, PhoneNumberSearchRequest searchRequest, Duration pollInterval) {
+        Objects.requireNonNull(countryCode, "'countryCode' cannot be null.");
+        Objects.requireNonNull(searchRequest, "'searchRequest' cannot be null.");
+        
+        if (pollInterval == null) {
+            pollInterval = defaultPollInterval;
+        }
+
+        return new PollerFlux<>(pollInterval,
+            searchAvailableNumbersInitOperation(countryCode, searchRequest),
+            searchAvailableNumbersPollOperation(),
+            cancelOperation(),
+            fetchFinalResultOperation());
+    }
+
+    private Function<PollingContext<PhoneNumberOperationResult>, Mono<PhoneNumberOperationResult>>
+        searchAvailableNumbersInitOperation(String countryCode, PhoneNumberSearchRequest searchRequest) {
+        return (pollingContext) -> {
+            return client.searchAvailablePhoneNumbersWithResponseAsync(countryCode, searchRequest)
+                .flatMap((PhoneNumbersSearchAvailablePhoneNumbersResponse response) -> {
+                    pollingContext.setData("operationId", parseIdFromUrl(response.getDeserializedHeaders().getOperationLocation()));
+                    pollingContext.setData("searchId", parseIdFromUrl(response.getDeserializedHeaders().getLocation()));
+                    return client.getOperationAsync(pollingContext.getData("operationId"))
+                    .flatMap((PhoneNumberOperationResult result) -> {
+                        return Mono.just(result);
+                    });
+                });
+        };
+    }
+
+    private String parseIdFromUrl(String url) {
+        String[] items = url.split("/");
+        String id = items[items.length - 1];
+        return id.substring(0, id.indexOf("?"));
```

Comment:
Not sure how Java handles a negative end index, is this safe?

Reply:
I will double check on this thanks